<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.552">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Kynon J Benjamin">
<meta name="dcterms.date" content="2024-03-30">

<title>Dimensional reduction with PHATE: Pericytes</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="01.pericyte_phate_files/libs/clipboard/clipboard.min.js"></script>
<script src="01.pericyte_phate_files/libs/quarto-html/quarto.js"></script>
<script src="01.pericyte_phate_files/libs/quarto-html/popper.min.js"></script>
<script src="01.pericyte_phate_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01.pericyte_phate_files/libs/quarto-html/anchor.min.js"></script>
<link href="01.pericyte_phate_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01.pericyte_phate_files/libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01.pericyte_phate_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01.pericyte_phate_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01.pericyte_phate_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#import-libraries" id="toc-import-libraries" class="nav-link active" data-scroll-target="#import-libraries"><span class="header-section-number">1</span> Import libraries</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data"><span class="header-section-number">2</span> Load data</a></li>
  <li><a href="#preprocessing-filtering-normalizing-and-transforming" id="toc-preprocessing-filtering-normalizing-and-transforming" class="nav-link" data-scroll-target="#preprocessing-filtering-normalizing-and-transforming"><span class="header-section-number">3</span> Preprocessing: Filtering, Normalizing, and Transforming</a>
  <ul class="collapse">
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering"><span class="header-section-number">3.1</span> Filtering</a></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization"><span class="header-section-number">3.2</span> Normalization</a></li>
  <li><a href="#transformation" id="toc-transformation" class="nav-link" data-scroll-target="#transformation"><span class="header-section-number">3.3</span> Transformation</a></li>
  </ul></li>
  <li><a href="#phate" id="toc-phate" class="nav-link" data-scroll-target="#phate"><span class="header-section-number">4</span> PHATE</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting"><span class="header-section-number">5</span> Plotting</a>
  <ul class="collapse">
  <li><a href="#angiotensin-ii-receptor-1-agtr1" id="toc-angiotensin-ii-receptor-1-agtr1" class="nav-link" data-scroll-target="#angiotensin-ii-receptor-1-agtr1"><span class="header-section-number">5.1</span> Angiotensin II receptor 1 (AGTR1)</a></li>
  <li><a href="#kde-plot-of-pericytes" id="toc-kde-plot-of-pericytes" class="nav-link" data-scroll-target="#kde-plot-of-pericytes"><span class="header-section-number">5.2</span> KDE plot of pericytes</a></li>
  <li><a href="#clustering-data-using-phate" id="toc-clustering-data-using-phate" class="nav-link" data-scroll-target="#clustering-data-using-phate"><span class="header-section-number">5.3</span> Clustering data using PHATE</a></li>
  </ul></li>
  <li><a href="#comparison-of-phate-clusters-and-agtr1-expressoin" id="toc-comparison-of-phate-clusters-and-agtr1-expressoin" class="nav-link" data-scroll-target="#comparison-of-phate-clusters-and-agtr1-expressoin"><span class="header-section-number">6</span> Comparison of PHATE clusters and AGTR1 expressoin</a>
  <ul class="collapse">
  <li><a href="#subset-data" id="toc-subset-data" class="nav-link" data-scroll-target="#subset-data"><span class="header-section-number">6.1</span> Subset data</a></li>
  <li><a href="#qualitative-review" id="toc-qualitative-review" class="nav-link" data-scroll-target="#qualitative-review"><span class="header-section-number">6.2</span> Qualitative review</a></li>
  <li><a href="#statistical-comparison" id="toc-statistical-comparison" class="nav-link" data-scroll-target="#statistical-comparison"><span class="header-section-number">6.3</span> Statistical comparison</a></li>
  </ul></li>
  <li><a href="#trajectory-inference" id="toc-trajectory-inference" class="nav-link" data-scroll-target="#trajectory-inference"><span class="header-section-number">7</span> Trajectory inference</a>
  <ul class="collapse">
  <li><a href="#normalize-based-counts" id="toc-normalize-based-counts" class="nav-link" data-scroll-target="#normalize-based-counts"><span class="header-section-number">7.1</span> Normalize based counts</a></li>
  <li><a href="#processing-for-paga" id="toc-processing-for-paga" class="nav-link" data-scroll-target="#processing-for-paga"><span class="header-section-number">7.2</span> Processing for PAGA</a></li>
  <li><a href="#clustering-and-paga" id="toc-clustering-and-paga" class="nav-link" data-scroll-target="#clustering-and-paga"><span class="header-section-number">7.3</span> Clustering and PAGA</a></li>
  <li><a href="#paga-embedding" id="toc-paga-embedding" class="nav-link" data-scroll-target="#paga-embedding"><span class="header-section-number">7.4</span> PAGA embedding</a></li>
  </ul></li>
  <li><a href="#pseudotime-visualization" id="toc-pseudotime-visualization" class="nav-link" data-scroll-target="#pseudotime-visualization"><span class="header-section-number">8</span> Pseudotime visualization</a>
  <ul class="collapse">
  <li><a href="#choose-a-root-cell-for-diffusion-pseudotime" id="toc-choose-a-root-cell-for-diffusion-pseudotime" class="nav-link" data-scroll-target="#choose-a-root-cell-for-diffusion-pseudotime"><span class="header-section-number">8.1</span> Choose a root cell for diffusion pseudotime</a></li>
  <li><a href="#compute-diffusion-pseudotime" id="toc-compute-diffusion-pseudotime" class="nav-link" data-scroll-target="#compute-diffusion-pseudotime"><span class="header-section-number">8.2</span> Compute diffusion pseudotime</a></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information"><span class="header-section-number">9</span> Session information</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="pericyte_phate.ipynb" download="pericyte_phate.ipynb"><i class="bi bi-journal-code"></i>Jupyter</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Dimensional reduction with PHATE: Pericytes</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Kynon J Benjamin </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 30, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>

<nav id="TOC-body" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#import-libraries" id="toc-import-libraries"><span class="header-section-number">1</span> Import libraries</a></li>
  <li><a href="#load-data" id="toc-load-data"><span class="header-section-number">2</span> Load data</a></li>
  <li><a href="#preprocessing-filtering-normalizing-and-transforming" id="toc-preprocessing-filtering-normalizing-and-transforming"><span class="header-section-number">3</span> Preprocessing: Filtering, Normalizing, and Transforming</a>
  <ul>
  <li><a href="#filtering" id="toc-filtering"><span class="header-section-number">3.1</span> Filtering</a></li>
  <li><a href="#normalization" id="toc-normalization"><span class="header-section-number">3.2</span> Normalization</a></li>
  <li><a href="#transformation" id="toc-transformation"><span class="header-section-number">3.3</span> Transformation</a></li>
  </ul></li>
  <li><a href="#phate" id="toc-phate"><span class="header-section-number">4</span> PHATE</a></li>
  <li><a href="#plotting" id="toc-plotting"><span class="header-section-number">5</span> Plotting</a>
  <ul>
  <li><a href="#angiotensin-ii-receptor-1-agtr1" id="toc-angiotensin-ii-receptor-1-agtr1"><span class="header-section-number">5.1</span> Angiotensin II receptor 1 (AGTR1)</a></li>
  <li><a href="#kde-plot-of-pericytes" id="toc-kde-plot-of-pericytes"><span class="header-section-number">5.2</span> KDE plot of pericytes</a></li>
  <li><a href="#clustering-data-using-phate" id="toc-clustering-data-using-phate"><span class="header-section-number">5.3</span> Clustering data using PHATE</a></li>
  </ul></li>
  <li><a href="#comparison-of-phate-clusters-and-agtr1-expressoin" id="toc-comparison-of-phate-clusters-and-agtr1-expressoin"><span class="header-section-number">6</span> Comparison of PHATE clusters and AGTR1 expressoin</a>
  <ul>
  <li><a href="#subset-data" id="toc-subset-data"><span class="header-section-number">6.1</span> Subset data</a></li>
  <li><a href="#qualitative-review" id="toc-qualitative-review"><span class="header-section-number">6.2</span> Qualitative review</a></li>
  <li><a href="#statistical-comparison" id="toc-statistical-comparison"><span class="header-section-number">6.3</span> Statistical comparison</a></li>
  </ul></li>
  <li><a href="#trajectory-inference" id="toc-trajectory-inference"><span class="header-section-number">7</span> Trajectory inference</a>
  <ul>
  <li><a href="#normalize-based-counts" id="toc-normalize-based-counts"><span class="header-section-number">7.1</span> Normalize based counts</a></li>
  <li><a href="#processing-for-paga" id="toc-processing-for-paga"><span class="header-section-number">7.2</span> Processing for PAGA</a></li>
  <li><a href="#clustering-and-paga" id="toc-clustering-and-paga"><span class="header-section-number">7.3</span> Clustering and PAGA</a></li>
  <li><a href="#paga-embedding" id="toc-paga-embedding"><span class="header-section-number">7.4</span> PAGA embedding</a></li>
  </ul></li>
  <li><a href="#pseudotime-visualization" id="toc-pseudotime-visualization"><span class="header-section-number">8</span> Pseudotime visualization</a>
  <ul>
  <li><a href="#choose-a-root-cell-for-diffusion-pseudotime" id="toc-choose-a-root-cell-for-diffusion-pseudotime"><span class="header-section-number">8.1</span> Choose a root cell for diffusion pseudotime</a></li>
  <li><a href="#compute-diffusion-pseudotime" id="toc-compute-diffusion-pseudotime"><span class="header-section-number">8.2</span> Compute diffusion pseudotime</a></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information"><span class="header-section-number">9</span> Session information</a></li>
  </ul>
</nav>
<section id="import-libraries" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Import libraries</h1>
<div id="cf61d7e0-a0b3-4ef2-9721-139bc5724e87" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> polars <span class="im">as</span> pl</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.stats <span class="im">import</span> f_oneway</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> phate, scprep, session_info</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statsmodels.stats.multicomp <span class="im">import</span> MultiComparison</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c69f6b63-ca18-4a9b-bc20-23ce9f242d26" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>matplotlib inline</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Load data</h1>
<div id="38a53dd0-58e1-4056-a980-4d951272495b" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"../_m/peri.hlca_core.dataset.h5ad"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>AnnData object with n_obs × n_vars = 3032 × 28024
    obs: 'suspension_type', 'donor_id', 'is_primary_data', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'tissue_ontology_term_id', 'organism_ontology_term_id', 'sex_ontology_term_id', 'BMI', 'age_or_mean_of_age_range', 'age_range', 'anatomical_region_ccf_score', 'ann_coarse_for_GWAS_and_modeling', 'ann_finest_level', 'ann_level_1', 'ann_level_2', 'ann_level_3', 'ann_level_4', 'ann_level_5', 'cause_of_death', 'dataset', 'entropy_dataset_leiden_3', 'entropy_original_ann_level_1_leiden_3', 'entropy_original_ann_level_2_clean_leiden_3', 'entropy_original_ann_level_3_clean_leiden_3', 'entropy_subject_ID_leiden_3', 'fresh_or_frozen', 'leiden_1', 'leiden_2', 'leiden_3', 'leiden_4', 'leiden_5', 'log10_total_counts', 'lung_condition', 'mixed_ancestry', 'n_genes_detected', 'original_ann_highest_res', 'original_ann_level_1', 'original_ann_level_2', 'original_ann_level_3', 'original_ann_level_4', 'original_ann_level_5', 'original_ann_nonharmonized', 'reannotation_type', 'reference_genome', 'sample', 'scanvi_label', 'sequencing_platform', 'size_factors', 'smoking_status', 'study', 'subject_type', 'tissue_dissociation_protocol', 'tissue_level_2', 'tissue_level_3', 'tissue_sampling_method', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'subclusters', 'clusters', 'compartment', 'patient', 'label'
    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'
    uns: 'X_name', 'batch_condition', 'citation', 'default_embedding', 'schema_reference', 'schema_version', 'title'
    obsm: 'X_scanvi_emb', 'X_umap'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="5ec3863c-694d-4390-aa17-679e8b212699" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>hlca_counts <span class="op">=</span> pd.DataFrame(adata.X.toarray(), </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                          columns<span class="op">=</span>[adata.var.feature_name],</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                          index<span class="op">=</span>adata.obs.index)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(hlca_counts.shape)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>hlca_counts.head(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(3032, 28024)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">feature_name</th>
<th data-quarto-table-cell-role="th">TSPAN6</th>
<th data-quarto-table-cell-role="th">TNMD</th>
<th data-quarto-table-cell-role="th">DPM1</th>
<th data-quarto-table-cell-role="th">SCYL3</th>
<th data-quarto-table-cell-role="th">C1orf112</th>
<th data-quarto-table-cell-role="th">FGR</th>
<th data-quarto-table-cell-role="th">CFH</th>
<th data-quarto-table-cell-role="th">FUCA2</th>
<th data-quarto-table-cell-role="th">GCLC</th>
<th data-quarto-table-cell-role="th">NFYA</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">LINC02456</th>
<th data-quarto-table-cell-role="th">TRBV6-2</th>
<th data-quarto-table-cell-role="th">RP1-221C16.8</th>
<th data-quarto-table-cell-role="th">LBHD2</th>
<th data-quarto-table-cell-role="th">RP11-124N19.4</th>
<th data-quarto-table-cell-role="th">RP11-11M20.4</th>
<th data-quarto-table-cell-role="th">LLNLR-245B6.1</th>
<th data-quarto-table-cell-role="th">MGC4859</th>
<th data-quarto-table-cell-role="th">RP11-107E5.4</th>
<th data-quarto-table-cell-role="th">RP11-299P2.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GGGACCTAGGTCATCT_HD68</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>2.669188</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">TCGGATAAGGTGAGCT_SC86</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>1.256804</td>
<td>0.0</td>
<td>0.0</td>
<td>0.814075</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>2 rows × 28024 columns</p>
</div>
</div>
</div>
</div>
<div id="56ab57f2-1b20-4d54-bd26-b8ffce507d2d" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>scprep.plot.plot_library_size(hlca_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/scprep/plot/utils.py:104: UserWarning:

Matplotlib is currently using module://matplotlib_inline.backend_inline, which is a non-GUI backend, so cannot show the figure.
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-6-output-2.png" width="660" height="467" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="preprocessing-filtering-normalizing-and-transforming" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Preprocessing: Filtering, Normalizing, and Transforming</h1>
<section id="filtering" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="filtering"><span class="header-section-number">3.1</span> Filtering</h2>
<p>This data, however, has already had very extensive filtering and QC. Based on the core model, there are no high mitochondria percentages and few outliers. So, we will skip the outlier step as this eliminates very few cells (&lt; 1%).</p>
<div id="c039ae3e-a491-4476-983d-24e0df1b6658" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>hlca_counts <span class="op">=</span> scprep.<span class="bu">filter</span>.filter_rare_genes(hlca_counts, min_cells<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>hlca_counts.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>(3032, 14395)</code></pre>
</div>
</div>
</section>
<section id="normalization" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="normalization"><span class="header-section-number">3.2</span> Normalization</h2>
<div id="ce9d7054-ce86-487c-af31-9a603b71f56a" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>hlca_counts <span class="op">=</span> scprep.normalize.library_size_normalize(hlca_counts)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>hlca_counts.head(<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">feature_name</th>
<th data-quarto-table-cell-role="th">TSPAN6</th>
<th data-quarto-table-cell-role="th">DPM1</th>
<th data-quarto-table-cell-role="th">SCYL3</th>
<th data-quarto-table-cell-role="th">C1orf112</th>
<th data-quarto-table-cell-role="th">FGR</th>
<th data-quarto-table-cell-role="th">CFH</th>
<th data-quarto-table-cell-role="th">FUCA2</th>
<th data-quarto-table-cell-role="th">GCLC</th>
<th data-quarto-table-cell-role="th">NFYA</th>
<th data-quarto-table-cell-role="th">STPG1</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">RP11-295P9.13</th>
<th data-quarto-table-cell-role="th">RP11-422P24.12</th>
<th data-quarto-table-cell-role="th">CTD-3113P16.11</th>
<th data-quarto-table-cell-role="th">RP11-605B16.1</th>
<th data-quarto-table-cell-role="th">BISPR</th>
<th data-quarto-table-cell-role="th">CTC-490E21.14</th>
<th data-quarto-table-cell-role="th">PRNCR1</th>
<th data-quarto-table-cell-role="th">RP11-557N21.1</th>
<th data-quarto-table-cell-role="th">RP11-11M20.4</th>
<th data-quarto-table-cell-role="th">LLNLR-245B6.1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GGGACCTAGGTCATCT_HD68</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>20.325068</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">TCGGATAAGGTGAGCT_SC86</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>6.653818</td>
<td>0.0</td>
<td>0.0</td>
<td>4.309905</td>
<td>0.0</td>
<td>...</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>2 rows × 14395 columns</p>
</div>
</div>
</div>
</div>
</section>
<section id="transformation" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="transformation"><span class="header-section-number">3.3</span> Transformation</h2>
<div id="f46aeac5-5418-426c-ac25-0b6ef2710f38" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>hlca_counts <span class="op">=</span> scprep.transform.sqrt(hlca_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="phate" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> PHATE</h1>
<div id="74c2a50c-a475-4979-990f-1d44287f8be5" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>phate_operator <span class="op">=</span> phate.PHATE(n_jobs<span class="op">=-</span><span class="dv">2</span>, random_state<span class="op">=</span><span class="dv">13</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>Y_phate <span class="op">=</span> phate_operator.fit_transform(hlca_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Calculating PHATE...
  Running PHATE on 3032 observations and 14395 variables.
  Calculating graph and diffusion operator...
    Calculating PCA...
    Calculated PCA in 9.93 seconds.
    Calculating KNN search...
    Calculated KNN search in 1.91 seconds.
    Calculating affinities...
    Calculated affinities in 0.59 seconds.
  Calculated graph and diffusion operator in 12.60 seconds.
  Calculating landmark operator...
    Calculating SVD...
    Calculated SVD in 0.54 seconds.
    Calculating KMeans...
    Calculated KMeans in 6.72 seconds.
  Calculated landmark operator in 8.94 seconds.
  Calculating optimal t...
    Automatically selected t = 12
  Calculated optimal t in 5.87 seconds.
  Calculating diffusion potential...
  Calculated diffusion potential in 2.08 seconds.
  Calculating metric MDS...
  Calculated metric MDS in 7.28 seconds.
Calculated PHATE in 36.78 seconds.</code></pre>
</div>
</div>
<div id="f48ee2bb-1240-4d34-af4d-2b326612ee73" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>scprep.plot.scatter2d(Y_phate, ticks<span class="op">=</span><span class="va">None</span>, label_prefix<span class="op">=</span><span class="st">"PHATE"</span>, figsize<span class="op">=</span>(<span class="dv">5</span>,<span class="dv">5</span>),</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                      filename<span class="op">=</span><span class="st">"pericyte_phate_clustering.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/scprep/plot/utils.py:104: UserWarning:

Matplotlib is currently using module://matplotlib_inline.backend_inline, which is a non-GUI backend, so cannot show the figure.
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-11-output-2.png" width="469" height="469" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="plotting" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Plotting</h1>
<section id="angiotensin-ii-receptor-1-agtr1" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="angiotensin-ii-receptor-1-agtr1"><span class="header-section-number">5.1</span> Angiotensin II receptor 1 (AGTR1)</h2>
<div id="979bc3cf-2026-4add-a455-eed14f6bf776" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plotting mitochondrial expression</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">4</span>,<span class="dv">4</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>scprep.plot.scatter2d(Y_phate, ax<span class="op">=</span>ax, c<span class="op">=</span>hlca_counts[<span class="st">'AGTR1'</span>],</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                      title<span class="op">=</span><span class="st">'AGTR1'</span>, ticks<span class="op">=</span><span class="va">False</span>, label_prefix<span class="op">=</span><span class="st">'PHATE'</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                      filename<span class="op">=</span><span class="st">"pericyte_phate_clustering.AGTR1.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-12-output-1.png" width="359" height="373" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="kde-plot-of-pericytes" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="kde-plot-of-pericytes"><span class="header-section-number">5.2</span> KDE plot of pericytes</h2>
<div id="1de0fa7c-faf5-49f7-a09d-03eb92d5ef57" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(<span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">4</span>,<span class="dv">4</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>kde_plot <span class="op">=</span> sns.kdeplot(x<span class="op">=</span>Y_phate[:, <span class="dv">0</span>], y<span class="op">=</span>Y_phate[:, <span class="dv">1</span>], fill<span class="op">=</span><span class="va">True</span>, ax<span class="op">=</span>ax, zorder<span class="op">=</span><span class="dv">0</span>, </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                       n_levels<span class="op">=</span><span class="dv">100</span>, cmap<span class="op">=</span><span class="st">"inferno"</span>)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>ax.set_xticks([])<span class="op">;</span> ax.set_yticks([])</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'PHATE 1'</span>, fontsize<span class="op">=</span><span class="dv">18</span>)<span class="op">;</span> ax.set_ylabel(<span class="st">'PHATE 2'</span>, fontsize<span class="op">=</span><span class="dv">18</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>ax.set_title(<span class="st">'KDE - Pericytes'</span>, fontsize<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> kde_plot.get_figure()</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"pericyte_phate_kde.pdf"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-13-output-1.png" width="374" height="374" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="clustering-data-using-phate" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="clustering-data-using-phate"><span class="header-section-number">5.3</span> Clustering data using PHATE</h2>
<div id="cd5e26a7-b973-42d4-87dd-5638447891e7" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>clusters <span class="op">=</span> phate.cluster.kmeans(phate_operator, k<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>scprep.plot.scatter2d(Y_phate, c<span class="op">=</span>clusters, s<span class="op">=</span><span class="dv">1</span>,figsize<span class="op">=</span>(<span class="fl">4.3</span>,<span class="dv">4</span>), ticks<span class="op">=</span><span class="va">None</span>, label_prefix<span class="op">=</span><span class="st">'PHATE'</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                      legend_anchor<span class="op">=</span>(<span class="dv">1</span>,<span class="dv">1</span>), fontsize<span class="op">=</span><span class="dv">12</span>, title<span class="op">=</span><span class="st">'PHATE clusters'</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                      filename<span class="op">=</span><span class="st">"pericyte_phate_clustering.kmeans.pdf"</span>, dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/phate/cluster.py:57: FutureWarning:

k is deprecated. Please use n_clusters in future.

/users/jbenjami/.local/lib/python3.9/site-packages/scprep/plot/utils.py:104: UserWarning:

Matplotlib is currently using module://matplotlib_inline.backend_inline, which is a non-GUI backend, so cannot show the figure.
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-14-output-2.png" width="397" height="367" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="comparison-of-phate-clusters-and-agtr1-expressoin" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> Comparison of PHATE clusters and AGTR1 expressoin</h1>
<section id="subset-data" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="subset-data"><span class="header-section-number">6.1</span> Subset data</h2>
<div id="8159d4ae-d3d9-4f2d-bcf9-7a7f15c47aeb" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>gene_list <span class="op">=</span> []</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> item <span class="kw">in</span> hlca_counts.columns:</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    gene_list.append(<span class="bu">str</span>(item).replace(<span class="st">"('"</span>, <span class="st">""</span>).replace(<span class="st">"',)"</span>, <span class="st">""</span>))</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[hlca_counts.index, adata.var.feature_name.isin(gene_list)]</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">"X_phate"</span>] <span class="op">=</span> Y_phate</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"phate_clusters"</span>] <span class="op">=</span> [<span class="st">"Group_"</span><span class="op">+</span><span class="bu">str</span>(clu) <span class="cf">for</span> clu <span class="kw">in</span> clusters]</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_98817/2217928001.py:6: ImplicitModificationWarning:

Setting element `.obsm['X_phate']` of view, initializing view as actual.
</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>AnnData object with n_obs × n_vars = 3032 × 14395
    obs: 'suspension_type', 'donor_id', 'is_primary_data', 'assay_ontology_term_id', 'cell_type_ontology_term_id', 'development_stage_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'tissue_ontology_term_id', 'organism_ontology_term_id', 'sex_ontology_term_id', 'BMI', 'age_or_mean_of_age_range', 'age_range', 'anatomical_region_ccf_score', 'ann_coarse_for_GWAS_and_modeling', 'ann_finest_level', 'ann_level_1', 'ann_level_2', 'ann_level_3', 'ann_level_4', 'ann_level_5', 'cause_of_death', 'dataset', 'entropy_dataset_leiden_3', 'entropy_original_ann_level_1_leiden_3', 'entropy_original_ann_level_2_clean_leiden_3', 'entropy_original_ann_level_3_clean_leiden_3', 'entropy_subject_ID_leiden_3', 'fresh_or_frozen', 'leiden_1', 'leiden_2', 'leiden_3', 'leiden_4', 'leiden_5', 'log10_total_counts', 'lung_condition', 'mixed_ancestry', 'n_genes_detected', 'original_ann_highest_res', 'original_ann_level_1', 'original_ann_level_2', 'original_ann_level_3', 'original_ann_level_4', 'original_ann_level_5', 'original_ann_nonharmonized', 'reannotation_type', 'reference_genome', 'sample', 'scanvi_label', 'sequencing_platform', 'size_factors', 'smoking_status', 'study', 'subject_type', 'tissue_dissociation_protocol', 'tissue_level_2', 'tissue_level_3', 'tissue_sampling_method', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid', 'subclusters', 'clusters', 'compartment', 'patient', 'label', 'phate_clusters'
    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'
    uns: 'X_name', 'batch_condition', 'citation', 'default_embedding', 'schema_reference', 'schema_version', 'title'
    obsm: 'X_scanvi_emb', 'X_umap', 'X_phate'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
</section>
<section id="qualitative-review" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="qualitative-review"><span class="header-section-number">6.2</span> Qualitative review</h2>
<div id="a96cc062-2bde-42de-a265-aa86639034fe" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>hlca_counts.loc[:, <span class="st">"PHATE"</span>] <span class="op">=</span> clusters</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>hlca_counts.loc[:, <span class="st">"patient"</span>] <span class="op">=</span> adata.obs[<span class="st">"donor_id"</span>]</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pl.from_pandas(hlca_counts.loc[:, [<span class="st">"AGTR1"</span>, <span class="st">"PHATE"</span>, <span class="st">"patient"</span>]])<span class="op">\</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>       .rename({<span class="st">"('AGTR1',)"</span>: <span class="st">"AGTR1"</span>, <span class="st">"('PHATE',)"</span>: <span class="st">"PHATE"</span>, </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                <span class="st">"('patient',)"</span>: <span class="st">"patient"</span>})</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>df.group_by([<span class="st">"PHATE"</span>, <span class="st">"patient"</span>]).agg([pl.<span class="bu">sum</span>(<span class="st">"AGTR1"</span>).name.suffix(<span class="st">"_sum"</span>), </span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                                       pl.mean(<span class="st">"AGTR1"</span>).name.suffix(<span class="st">"_mean"</span>),</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>                                       pl.count(<span class="st">"AGTR1"</span>).name.suffix(<span class="st">"_count"</span>),</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>                                       pl.std(<span class="st">"AGTR1"</span>).name.suffix(<span class="st">"_std"</span>)])<span class="op">\</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  .sort(<span class="st">"PHATE"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>
<div><style>
.dataframe > thead > tr > th,
.dataframe > tbody > tr > td {
  text-align: right;
  white-space: pre-wrap;
}
</style>
<small>shape: (54, 6)</small>
<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th">PHATE</th>
<th data-quarto-table-cell-role="th">patient</th>
<th data-quarto-table-cell-role="th">AGTR1_sum</th>
<th data-quarto-table-cell-role="th">AGTR1_mean</th>
<th data-quarto-table-cell-role="th">AGTR1_count</th>
<th data-quarto-table-cell-role="th">AGTR1_std</th>
</tr>
<tr class="odd">
<th>i32</th>
<th>cat</th>
<th>f64</th>
<th>f64</th>
<th>u32</th>
<th>f64</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>2.610938</td>
<td>0.217578</td>
<td>12</td>
<td>0.753713</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>12.07757</td>
<td>1.207757</td>
<td>10</td>
<td>1.588502</td>
</tr>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>5</td>
<td>0.0</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>4.371114</td>
<td>4.371114</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>3.586837</td>
<td>1.793418</td>
<td>2</td>
<td>0.189102</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>2.893894</td>
<td>2.893894</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>1740.951871</td>
<td>1.544767</td>
<td>1127</td>
<td>1.420807</td>
</tr>
<tr class="odd">
<td>0</td>
<td>"homosapiens_No…</td>
<td>9.064326</td>
<td>1.133041</td>
<td>8</td>
<td>2.135499</td>
</tr>
<tr class="even">
<td>0</td>
<td>"homosapiens_No…</td>
<td>90.379624</td>
<td>1.481633</td>
<td>61</td>
<td>1.500975</td>
</tr>
<tr class="odd">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="even">
<td>2</td>
<td>"homosapiens_No…</td>
<td>183.989071</td>
<td>1.769126</td>
<td>104</td>
<td>1.673966</td>
</tr>
<tr class="odd">
<td>2</td>
<td>"homosapiens_No…</td>
<td>6.737752</td>
<td>1.684438</td>
<td>4</td>
<td>1.947152</td>
</tr>
<tr class="even">
<td>2</td>
<td>"homosapiens_No…</td>
<td>66.092722</td>
<td>1.201686</td>
<td>55</td>
<td>1.646414</td>
</tr>
<tr class="odd">
<td>2</td>
<td>"homosapiens_No…</td>
<td>40.693629</td>
<td>2.14177</td>
<td>19</td>
<td>1.560494</td>
</tr>
<tr class="even">
<td>2</td>
<td>"homosapiens_No…</td>
<td>31.99676</td>
<td>2.133117</td>
<td>15</td>
<td>1.404701</td>
</tr>
<tr class="odd">
<td>2</td>
<td>"homosapiens_No…</td>
<td>7.166006</td>
<td>1.194334</td>
<td>6</td>
<td>1.881717</td>
</tr>
<tr class="even">
<td>2</td>
<td>"homosapiens_No…</td>
<td>6.016865</td>
<td>2.005622</td>
<td>3</td>
<td>1.768744</td>
</tr>
<tr class="odd">
<td>2</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="even">
<td>2</td>
<td>"homosapiens_No…</td>
<td>7.239764</td>
<td>2.413255</td>
<td>3</td>
<td>0.502074</td>
</tr>
<tr class="odd">
<td>2</td>
<td>"homosapiens_No…</td>
<td>0.0</td>
<td>0.0</td>
<td>1</td>
<td>null</td>
</tr>
<tr class="even">
<td>3</td>
<td>"homosapiens_No…</td>
<td>692.243147</td>
<td>1.784132</td>
<td>388</td>
<td>1.440053</td>
</tr>
<tr class="odd">
<td>3</td>
<td>"homosapiens_No…</td>
<td>3.21802</td>
<td>1.60901</td>
<td>2</td>
<td>2.275484</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="f7e8228f-86de-4521-8f37-25acd0b84143" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>dx <span class="op">=</span> df.group_by([<span class="st">"PHATE"</span>, <span class="st">"patient"</span>]).agg(pl.mean(<span class="st">"AGTR1"</span>))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.boxplot(data<span class="op">=</span>dx, x<span class="op">=</span><span class="st">"PHATE"</span>, y<span class="op">=</span><span class="st">"AGTR1"</span>, fill<span class="op">=</span><span class="va">False</span>, width<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>box_plot <span class="op">=</span> sns.stripplot(data<span class="op">=</span>dx, x<span class="op">=</span><span class="st">"PHATE"</span>, y<span class="op">=</span><span class="st">"AGTR1"</span>, ax<span class="op">=</span>ax)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">'PHATE Clusters'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)<span class="op">;</span> </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">'Normalized Expression (AGTR1)'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>fig.tight_layout()</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> box_plot.get_figure()</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>fig.savefig(<span class="st">"pericyte_phate.boxplot_AGTR1.pdf"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-17-output-1.png" width="579" height="432" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="statistical-comparison" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="statistical-comparison"><span class="header-section-number">6.3</span> Statistical comparison</h2>
<div id="fc9e6626-f3a0-4681-868a-7ba1ae8537bf" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>c0 <span class="op">=</span> dx.<span class="bu">filter</span>(pl.col(<span class="st">"PHATE"</span>) <span class="op">==</span> <span class="dv">0</span>).to_pandas().AGTR1.values</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>c1 <span class="op">=</span> dx.<span class="bu">filter</span>(pl.col(<span class="st">"PHATE"</span>) <span class="op">==</span> <span class="dv">1</span>).to_pandas().AGTR1.values</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>c2 <span class="op">=</span> dx.<span class="bu">filter</span>(pl.col(<span class="st">"PHATE"</span>) <span class="op">==</span> <span class="dv">2</span>).to_pandas().AGTR1.values</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>c3 <span class="op">=</span> dx.<span class="bu">filter</span>(pl.col(<span class="st">"PHATE"</span>) <span class="op">==</span> <span class="dv">3</span>).to_pandas().AGTR1.values</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>f_oneway(c0, c1, c2, c3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<pre><code>F_onewayResult(statistic=0.6774824662902309, pvalue=0.5699369537140733)</code></pre>
</div>
</div>
<div id="b1e5a653-64b0-4350-83e4-9d420eb38a47" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> MultiComparison(dx.to_pandas().AGTR1, dx.to_pandas().PHATE)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>res_mod <span class="op">=</span> model.tukeyhsd()</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(res_mod)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Multiple Comparison of Means - Tukey HSD, FWER=0.05
===================================================
group1 group2 meandiff p-adj   lower  upper  reject
---------------------------------------------------
     0      1  -0.2537 0.9349 -1.3985  0.891  False
     0      2  -0.4221 0.6321 -1.3599 0.5157  False
     0      3   0.3735 0.9695  -1.838  2.585  False
     1      2  -0.1684 0.9756 -1.2455 0.9088  False
     1      3   0.6273 0.8833 -1.6468 2.9014  False
     2      3   0.7956 0.7664 -1.3817 2.9729  False
---------------------------------------------------</code></pre>
</div>
</div>
</section>
</section>
<section id="trajectory-inference" class="level1" data-number="7">
<h1 data-number="7"><span class="header-section-number">7</span> Trajectory inference</h1>
<section id="normalize-based-counts" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="normalize-based-counts"><span class="header-section-number">7.1</span> Normalize based counts</h2>
<div id="5dc73fe4-d5d8-4a48-bd3d-44d9a3e9c1c4" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>adata.raw <span class="op">=</span> adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>normalizing counts per cell
    finished (0:00:00)</code></pre>
</div>
</div>
<div id="0ca2ebbe-041e-4178-9e32-75f22c8ad793" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>adata.var[adata.var.feature_name.isin([<span class="st">"AGTR1"</span>])]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">feature_is_filtered</th>
<th data-quarto-table-cell-role="th">feature_name</th>
<th data-quarto-table-cell-role="th">feature_reference</th>
<th data-quarto-table-cell-role="th">feature_biotype</th>
<th data-quarto-table-cell-role="th">feature_length</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000144891</td>
<td>False</td>
<td>AGTR1</td>
<td>NCBITaxon:9606</td>
<td>gene</td>
<td>3478</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</div>
</section>
<section id="processing-for-paga" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="processing-for-paga"><span class="header-section-number">7.2</span> Processing for PAGA</h2>
<div id="cb101385-8053-4bbf-b5b0-90d434860ce9" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata, n_neighbors<span class="op">=</span><span class="dv">50</span>, use_rep<span class="op">=</span><span class="st">"X_phate"</span>, random_state<span class="op">=</span><span class="dv">13</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>sc.tl.draw_graph(adata, random_state<span class="op">=</span><span class="dv">13</span>)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span>[<span class="st">"phate_clusters"</span>, <span class="st">"ENSG00000144891"</span>], </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                 legend_loc<span class="op">=</span><span class="st">"on data"</span>, save<span class="op">=</span><span class="st">".pericyte_phate.pdf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing neighbors
    finished: added to `.uns['neighbors']`
    `.obsp['distances']`, distances for each pair of neighbors
    `.obsp['connectivities']`, weighted adjacency matrix (0:00:15)
drawing single-cell graph using layout 'fa'
WARNING: Package 'fa2' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2' (`pip install fa2`).
    finished: added
    'X_draw_graph_fr', graph_drawing coordinates (adata.obsm) (0:00:23)
WARNING: You are using a low resolution (dpi&lt;150) for saving figures.
Consider running `set_figure_params(dpi_save=...)`, which will adjust `matplotlib.rcParams['savefig.dpi']`
WARNING: saving figure to file figures/draw_graph_fr.pericyte_phate.pdf</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning:

No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-22-output-3.png" width="1239" height="427" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="clustering-and-paga" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="clustering-and-paga"><span class="header-section-number">7.3</span> Clustering and PAGA</h2>
<div id="449c5385-ba91-4e83-b87d-96b37851c749" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, resolution<span class="op">=</span><span class="fl">0.5</span>, random_state<span class="op">=</span><span class="dv">13</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sc.tl.paga(adata, groups<span class="op">=</span><span class="st">"leiden"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>sc.pl.paga(adata, color<span class="op">=</span>[<span class="st">"leiden"</span>, <span class="st">"phate_clusters"</span>, <span class="st">"ENSG00000144891"</span>],</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>           save<span class="op">=</span><span class="st">".pericyte_phate.leiden.pdf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>running Leiden clustering
    finished: found 12 clusters and added
    'leiden', the cluster labels (adata.obs, categorical) (0:00:00)
running PAGA
    finished: added
    'paga/connectivities', connectivities adjacency (adata.uns)
    'paga/connectivities_tree', connectivities subtree (adata.uns) (0:00:00)
--&gt; added 'pos', the PAGA positions (adata.uns['paga'])
WARNING: saving figure to file figures/paga.pericyte_phate.leiden.pdf</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-23-output-2.png" width="2206" height="423" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="paga-embedding" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="paga-embedding"><span class="header-section-number">7.4</span> PAGA embedding</h2>
<div id="ba7f767d-bf64-4f0e-b9cb-356ef67cd4a9" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sc.tl.draw_graph(adata, init_pos<span class="op">=</span><span class="st">"paga"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span>[<span class="st">"leiden"</span>, <span class="st">"phate_clusters"</span>, <span class="st">"ENSG00000144891"</span>], </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                 legend_loc<span class="op">=</span><span class="st">"on data"</span>, save<span class="op">=</span><span class="st">".pericyte_phate.paga_embedding.pdf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>drawing single-cell graph using layout 'fa'
WARNING: Package 'fa2' is not installed, falling back to layout 'fr'.To use the faster and better ForceAtlas2 layout, install package 'fa2' (`pip install fa2`).
    finished: added
    'X_draw_graph_fr', graph_drawing coordinates (adata.obsm) (0:00:10)
WARNING: saving figure to file figures/draw_graph_fr.pericyte_phate.paga_embedding.pdf</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning:

No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored

/users/jbenjami/.local/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning:

No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-24-output-3.png" width="1846" height="427" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="pseudotime-visualization" class="level1" data-number="8">
<h1 data-number="8"><span class="header-section-number">8</span> Pseudotime visualization</h1>
<section id="choose-a-root-cell-for-diffusion-pseudotime" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="choose-a-root-cell-for-diffusion-pseudotime"><span class="header-section-number">8.1</span> Choose a root cell for diffusion pseudotime</h2>
<div id="83220302-a289-481f-82ba-7b1ee02ee048" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>adata.uns[<span class="st">"iroot"</span>] <span class="op">=</span> np.flatnonzero(adata.obs[<span class="st">"leiden"</span>] <span class="op">==</span> <span class="st">'11'</span>)[<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="compute-diffusion-pseudotime" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="compute-diffusion-pseudotime"><span class="header-section-number">8.2</span> Compute diffusion pseudotime</h2>
<div id="5b1243a3-7fb4-4b65-ba48-ba2c7937ee91" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>sc.tl.dpt(adata)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>sc.pl.draw_graph(adata, color<span class="op">=</span>[<span class="st">"phate_clusters"</span>, <span class="st">"ENSG00000144891"</span>, <span class="st">"dpt_pseudotime"</span>], </span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>                 legend_loc<span class="op">=</span><span class="st">"on data"</span>, save<span class="op">=</span><span class="st">".pericytes_phate.diff_pseudotime.pdf"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.
computing Diffusion Maps using n_comps=15(=n_dcs)
computing transitions
    finished (0:00:00)
    eigenvalues of transition matrix
    [1.         0.99956065 0.99860865 0.9981825  0.9964726  0.99345714
     0.99140364 0.9888171  0.9869463  0.9852987  0.98470664 0.98118716
     0.9786843  0.97644734 0.9713054 ]
    finished: added
    'X_diffmap', diffmap coordinates (adata.obsm)
    'diffmap_evals', eigenvalues of transition matrix (adata.uns) (0:00:00)
computing Diffusion Pseudotime using n_dcs=10
    finished: added
    'dpt_pseudotime', the pseudotime (adata.obs) (0:00:00)
WARNING: saving figure to file figures/draw_graph_fr.pericytes_phate.diff_pseudotime.pdf</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/users/jbenjami/.local/lib/python3.9/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning:

No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="01.pericyte_phate_files/figure-html/cell-26-output-3.png" width="1846" height="427" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="session-information" class="level1" data-number="9">
<h1 data-number="9"><span class="header-section-number">9</span> Session information</h1>
<div id="5f0f67b8-80dc-4a52-8613-fd7ace999b45" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>session_info.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<details>
<summary>Click to view session information</summary>
<pre>-----
anndata             0.10.5.post1
matplotlib          3.7.1
numpy               1.24.4
pandas              2.0.3
phate               1.0.11
polars              0.19.12
scanpy              1.9.8
scipy               1.8.1
scprep              1.2.3
seaborn             0.13.2
session_info        1.0.0
statsmodels         0.14.0
-----
</pre>
<details>
<summary>Click to view modules imported as dependencies</summary>
<pre>PIL                         8.2.0
anyio                       NA
argcomplete                 NA
arrow                       1.2.3
asttokens                   NA
attr                        23.2.0
attrs                       23.2.0
babel                       2.10.3
backcall                    0.2.0
beta_ufunc                  NA
binom_ufunc                 NA
cairo                       1.20.1
certifi                     2024.02.02
cffi                        1.14.5
chardet                     4.0.0
charset_normalizer          3.3.2
cloudpickle                 2.1.0
comm                        0.2.2
contourpy                   1.1.0
cycler                      0.10.0
cython_runtime              NA
dask                        2022.8.1
dateutil                    2.8.2
debugpy                     1.6.3
decorator                   5.1.1
deprecated                  1.2.13
exceptiongroup              1.0.0rc8
executing                   0.10.0
fastjsonschema              NA
fontTools                   4.33.3
fqdn                        NA
fsspec                      2022.7.1
future                      0.18.2
gi                          3.40.1
gio                         NA
glib                        NA
gobject                     NA
graphtools                  1.5.3
gtk                         NA
h5py                        3.7.0
hypergeom_ufunc             NA
idna                        2.10
igraph                      0.10.4
importlib_metadata          NA
importlib_resources         NA
ipykernel                   6.15.1
ipywidgets                  8.1.2
isoduration                 NA
jedi                        0.18.1
jinja2                      3.1.1
joblib                      1.2.0
json5                       NA
jsonpointer                 2.3
jsonschema                  4.21.1
jsonschema_specifications   NA
jupyter_events              0.10.0
jupyter_server              2.13.0
jupyterlab_server           2.25.4
kaleido                     0.2.1
kiwisolver                  1.4.3
leidenalg                   0.9.1
llvmlite                    0.42.0
louvain                     0.8.1
lxml                        4.9.1
markupsafe                  2.1.1
matplotlib_inline           0.1.6
mpl_toolkits                NA
natsort                     8.2.0
nbformat                    5.4.0
nbinom_ufunc                NA
netifaces                   0.10.6
networkx                    2.8.8
numba                       0.59.0
opt_einsum                  v3.3.0
overrides                   NA
packaging                   21.3
parso                       0.8.3
patsy                       0.5.2
pexpect                     4.8.0
pickleshare                 0.7.5
pkg_resources               NA
platformdirs                3.8.0
plotly                      5.9.0
ply                         3.11
prometheus_client           NA
prompt_toolkit              3.0.30
psutil                      5.9.1
ptyprocess                  0.7.0
pure_eval                   0.2.2
pyarrow                     15.0.0
pycparser                   2.20
pydev_ipython               NA
pydevconsole                NA
pydevd                      2.8.0
pydevd_file_utils           NA
pydevd_plugins              NA
pydevd_tracing              NA
pygments                    2.13.0
pygsp                       0.5.1
pynndescent                 0.5.8
pyparsing                   3.0.9
pythonjsonlogger            NA
pytz                        2022.1
referencing                 NA
requests                    2.31.0
rfc3339_validator           0.1.4
rfc3986_validator           0.1.1
rpds                        NA
s_gd2                       1.8.1
send2trash                  NA
six                         1.16.0
sklearn                     1.4.1.post1
sniffio                     1.2.0
socks                       1.7.1
stack_data                  0.4.0
swig_runtime_data4          NA
tasklogger                  1.2.0
tenacity                    NA
texttable                   1.6.7
threadpoolctl               3.1.0
tlz                         0.12.0
toolz                       0.12.0
torch                       1.13.1+cu117
tornado                     6.2
tqdm                        4.64.0
traitlets                   5.9.0
typing_extensions           NA
umap                        0.5.5
uri_template                NA
urllib3                     1.26.5
wcwidth                     0.2.5
webcolors                   1.11.1
websocket                   1.4.0
wrapt                       1.16.0
yaml                        6.0
zipp                        NA
zmq                         25.0.0
zoneinfo                    NA
zstandard                   0.18.0
</pre>
</details> <!-- seems like this ends pre, so might as well be explicit -->
<pre>-----
IPython             8.4.0
jupyter_client      8.0.2
jupyter_core        5.2.0
jupyterlab          4.1.5
notebook            7.1.2
-----
Python 3.9.16 (main, Dec  8 2022, 00:00:00) [GCC 11.3.1 20221121 (Red Hat 11.3.1-4)]
Linux-5.14.0-284.11.1.el9_2.x86_64-x86_64-with-glibc2.34
-----
Session information updated at 2024-03-30 16:25
</pre>
</details>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>